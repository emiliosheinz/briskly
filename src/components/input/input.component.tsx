import React from 'react'

import { classNames } from '~/utils/css'

import type { InputProps } from './input.types'

export const Input = React.forwardRef<HTMLInputElement, InputProps>(
  function Input(props, ref) {
    const { label, error, id, disabled, ...otherProps } = props

    const errorClassNames = error
      ? 'border-error-700 focus:border-error-700 focus:ring-error-700'
      : 'border-primary-500 focus:border-primary-900 focus:ring-primary-900'

    return (
      <div
        aria-disabled={disabled}
        className={classNames('w-full', disabled ? 'opacity-50' : '')}
      >
        <label
          htmlFor={id}
          className='block text-sm font-medium capitalize text-primary-800'
        >
          {label}
        </label>
        <div className='mt-1 rounded-md shadow-sm'>
          <input
            id={id}
            ref={ref}
            type='text'
            disabled={disabled}
            data-testid='briskly-custom-input'
            className={classNames(
              'block w-full rounded-md bg-primary-50 pl-6 pr-10 text-base text-primary-900',
              errorClassNames,
            )}
            {...otherProps}
          />
        </div>
        <p className='mt-1 h-5 overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700'>
          {error}
        </p>
      </div>
    )
  },
)
