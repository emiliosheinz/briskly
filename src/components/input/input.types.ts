export type InputProps = {
  error?: string
  label: string
  id: string
} & Omit<
  React.DetailedHTMLProps<
    React.InputHTMLAttributes<HTMLInputElement>,
    HTMLInputElement
  >,
  'type' | 'id'
>
