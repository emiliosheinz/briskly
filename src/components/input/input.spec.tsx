import { render, screen } from '~/utils/tests/react-testing-library'
import { Input } from './input.component'

describe('Input Component', () => {
  it('should be correctly rendered', () => {
    const component = render(<Input id='mock' label='Label' />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should be correctly rendered when hasError', () => {
    const component = render(
      <Input id='mock' label='Label' error='Mocked Error' />,
    )

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should be correctly rendered when disabled', () => {
    const component = render(<Input disabled id='mock' label='Label' />)

    expect(component.asFragment()).toMatchSnapshot()
  })

  it('should show input label', () => {
    render(<Input id='mock' label='Label' />)

    const label = screen.getByText('Label')

    expect(label).toBeInTheDocument()
  })

  it('should show input error', () => {
    render(<Input id='mock' label='Label' error='Mocked Error' />)

    const error = screen.getByText('Mocked Error')

    expect(error).toBeInTheDocument()
  })

  it('should disable input when disabled is provided', () => {
    render(<Input disabled id='mock' label='Label' error='Mocked Error' />)

    const input = screen.getByTestId('briskly-custom-input')

    expect(input).toBeDisabled()
  })
})
