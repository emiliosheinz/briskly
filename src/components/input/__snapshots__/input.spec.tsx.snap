// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Component should be correctly rendered 1`] = `
<DocumentFragment>
  <div
    class="w-full max-w-xs"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <input
        class="block w-full rounded-md border-primary-500 bg-primary-50 pl-6 pr-10 text-base text-primary-900 focus:border-primary-900 focus:ring-primary-900"
        data-testid="briskly-custom-input"
        id="mock"
        type="text"
      />
    </div>
    <p
      class="mt-1 h-5 max-w-xs overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    />
  </div>
</DocumentFragment>
`;

exports[`Input Component should be correctly rendered when disabled 1`] = `
<DocumentFragment>
  <div
    aria-disabled="true"
    class="w-full max-w-xs opacity-50"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <input
        class="block w-full rounded-md border-primary-500 bg-primary-50 pl-6 pr-10 text-base text-primary-900 focus:border-primary-900 focus:ring-primary-900"
        data-testid="briskly-custom-input"
        disabled=""
        id="mock"
        type="text"
      />
    </div>
    <p
      class="mt-1 h-5 max-w-xs overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    />
  </div>
</DocumentFragment>
`;

exports[`Input Component should be correctly rendered when hasError 1`] = `
<DocumentFragment>
  <div
    class="w-full max-w-xs"
  >
    <label
      class="block text-sm font-medium capitalize text-primary-800"
      for="mock"
    >
      Label
    </label>
    <div
      class="mt-1 rounded-md shadow-sm"
    >
      <input
        class="block w-full rounded-md border-primary-500 bg-primary-50 pl-6 pr-10 text-base text-primary-900 focus:border-primary-900 focus:ring-primary-900 border-error-700 focus:border-error-700 focus:ring-error-700"
        data-testid="briskly-custom-input"
        id="mock"
        type="text"
      />
    </div>
    <p
      class="mt-1 h-5 max-w-xs overflow-hidden text-ellipsis whitespace-nowrap text-sm capitalize text-error-700"
    >
      Mocked Error
    </p>
  </div>
</DocumentFragment>
`;
